apps/vor/pkg/postgres/lessonplanStore.go

```diff
@@ -1,173 +1,152 @@
+	// Create all instance of repeating plans and save to db. This will make it easy to
+	// retrieve, modify, and attach metadata to individual instances of the plans down the road
+	if planInput.Repetition != nil {
+		// If nil, repetition_type will automatically be 0, since it has useZero tag.
+		planDetails.RepetitionType = planInput.Repetition.Type
+		planDetails.RepetitionEndDate = planInput.Repetition.EndDate
+
+		currentDate := planInput.Date
+		monthToAdd := 0
+		daysToAdd := 0
+		switch planInput.Repetition.Type {
+		case cLessonPlan.RepetitionDaily:
+			daysToAdd = 1
+		case cLessonPlan.RepetitionWeekly:
+			daysToAdd = 7
+		case cLessonPlan.RepetitionMonthly:
+			monthToAdd = 1
 		}
-
-		for _, file := range planInput.FileIds {
-			if file == "" {
-				continue
-			}
-
-			objFile := FileToLessonPlan{
-				LessonPlanId: obj.Id,
-				FileId:       file,
-			}
-
-			if err := tx.Insert(&objFile); err != nil {
-				return err
+		for {
+			currentDate = currentDate.AddDate(0, monthToAdd, daysToAdd)
+			if currentDate.After(planInput.Repetition.EndDate) {
+				break
 			}
+			plans = append(plans, LessonPlan{
+				Id:        uuid.New().String(),
+				Date:      &currentDate,
+				DetailsId: planDetails.Id,
+			})
 		}
+	}
 
-		if planInput.EndTime != nil {
-			rpObj = RepetitionPattern{
-				LessonPlanId: obj.Id,
-				EndTime:      *planInput.EndTime,
-			}
+	fileRelations := make([]FileToLessonPlan, len(planInput.FileIds))
+	for idx, file := range planInput.FileIds {
+		fileRelations[idx] = FileToLessonPlan{
+			LessonPlanDetailsId: planDetails.Id,
+			FileId:              file,
+		}
+	}
 
-			if err := tx.Insert(&rpObj); err != nil {
-				return err
+	if err := s.RunInTransaction(func(tx *pg.Tx) error {
+		if err := tx.Insert(&planDetails); err != nil {
+			return richErrors.Wrap(err, "failed to save lesson plan details")
+		}
+		if err := tx.Insert(&plans); err != nil {
+			return richErrors.Wrap(err, "failed to save lesson plan")
+		}
+		if len(fileRelations) > 0 {
+			if err := tx.Insert(&fileRelations); err != nil {
+				return richErrors.Wrap(err, "failed to save file relations")
 			}
 		}
-
 		return nil
-	})
+	}); err != nil {
 
-	if err != nil {
-		err = errors.Wrapf(err, "failed create lesson plan")
 		return nil, err
 	}
 
 	return &cLessonPlan.LessonPlan{
-		Id:          obj.Id,
-		Title:       obj.Title,
-		Description: obj.Description,
-		ClassId:     obj.ClassId,
+		Id:          planDetails.Id,
+		Title:       *planDetails.Title,
+		Description: *planDetails.Description,
+		ClassId:     planDetails.ClassId,
 	}, nil
 }
 
-func (s LessonPlanStore) UpdateLessonPlan(planData cLessonPlan.UpdatePlanData) (int, error) {
-	// TODO: need to be simplified
-	obj := LessonPlan{
-		Id: planData.PlanId,
+func (s LessonPlanStore) UpdateLessonPlan(planInput cLessonPlan.UpdatePlanData) (int, error) {
+	plan := LessonPlan{
+		Id:   planInput.Id,
+		Date: planInput.Date,
+	}
+	planDetails := LessonPlanDetails{
+		Title:       planInput.Title,
```

Comment:
shouldn't we need to check whether the data is empty or not first to prevent nil error? or go-pg already handle that

Reply:
not a pointer anymore now. But yes, go-pg would handle nil values just fine, it'll become NULL when it got inserted to postgres.